query/src/main/scala/filodb/query/exec/rangefn/RateFunctions.scala

```diff
@@ -100,3 +165,74 @@ object DeltaFunction extends RangeFunction {
     sampleToEmit.setValues(endTimestamp, result.getOrElse(Double.NaN)) // TODO need to use a NA instead of NaN
   }
 }
+
+object IRateFunction extends RangeFunction {
+
+  override def needsCounterCorrection: Boolean = true
+  def addToWindow(row: TransientRow): Unit = {}
+  def removeFromWindow(row: TransientRow): Unit = {}
+
+  def apply(startTimestamp: Long,
+            endTimestamp: Long,
+            window: Window,
+            sampleToEmit: TransientRow,
+            queryConfig: QueryConfig): Unit = {
+    val result = RateFunctions.instantValue(startTimestamp,
+      endTimestamp, window, true)
+    sampleToEmit.setValues(endTimestamp, result.getOrElse(Double.NaN)) // TODO need to use a NA instead of NaN
+  }
+}
+
+object IDeltaFunction extends RangeFunction {
+
+  def addToWindow(row: TransientRow): Unit = {}
+  def removeFromWindow(row: TransientRow): Unit = {}
+
+  def apply(startTimestamp: Long,
+            endTimestamp: Long,
+            window: Window,
+            sampleToEmit: TransientRow,
+            queryConfig: QueryConfig): Unit = {
+    val result = RateFunctions.instantValue(startTimestamp,
+      endTimestamp, window, false)
+    sampleToEmit.setValues(endTimestamp, result.getOrElse(Double.NaN)) // TODO need to use a NA instead of NaN
+  }
+}
+
+object DerivFunction extends RangeFunction {
+
+  def addToWindow(row: TransientRow): Unit = {}
+  def removeFromWindow(row: TransientRow): Unit = {}
+
+  def apply(startTimestamp: Long,
+            endTimestamp: Long,
+            window: Window,
+            sampleToEmit: TransientRow,
+            queryConfig: QueryConfig): Unit = {
+    val result = RateFunctions.derivFunction(window)
+    sampleToEmit.setValues(endTimestamp, result.getOrElse(Double.NaN)) // TODO need to use a NA instead of NaN
+  }
+}
+
+object ResetsFunction extends RangeFunction {
+
+  def addToWindow(row: TransientRow): Unit = {}
+  def removeFromWindow(row: TransientRow): Unit = {}
+
+  def apply(startTimestamp: Long,
+            endTimestamp: Long,
+            window: Window,
+            sampleToEmit: TransientRow,
+            queryConfig: QueryConfig): Unit = {
+    var resets = 0
+    var prev = window.head.value
+    for (i <- 1 until window.size) {
```

Comment:
This can be done more efficiently with O(n) complexity using the addToWindow and removeFromWindow functions. This implementation will result in O(n^w) complexity, where w is length of time window

Reply:
👍 on it